lapce-data/src/hover.rs

```diff
@@ -152,41 +170,317 @@ impl Default for HoverData {
+        }
+        layout.draw(ctx, origin);
+    }
+}
+// Let this be treated as Text
+impl PietTextStorage for MarkdownText {
+    fn as_str(&self) -> &str {
+        self.text.as_str()
+    }
+}
+impl TextStorage for MarkdownText {
+    fn add_attributes(
+        &self,
+        builder: druid::piet::PietTextLayoutBuilder,
+        env: &druid::Env,
+    ) -> druid::piet::PietTextLayoutBuilder {
+        self.text.add_attributes(builder, env)
+    }
+
+    fn env_update(&self, ctx: &druid::text::EnvUpdateCtx) -> bool {
+        self.text.env_update(ctx)
+    }
+
+    fn links(&self) -> &[druid::text::Link] {
+        self.text.links()
+    }
+}
+
+// TODO: This could moved into a more general utility file, for use in things other than hover
+/// Parse the given markdown into renderable druid rich text
+/// `base_font_size` is the font size of normal text, and the size of headers is computed from it
+fn parse_markdown(text: &str, style: HoverTextStyle) -> MarkdownText {
+    use pulldown_cmark::{Event, Options, Parser};
+
+    let mut builder = RichTextBuilder::new();
+
+    // Our position within the text
+    let mut pos = 0;
+
+    // TODO: (minor): This could use a smallvec since most tags are probably not that nested
+    // Stores the current tags (like italics/bold/strikethrough) so that they can be nested
+    let mut tag_stack = Vec::new();
+
+    let mut code_block_indices = Vec::new();
+
+    // Construct the markdown parser. We enable most of the options in order to provide the most
+    // compatibility that pulldown_cmark allows.
+    let parser = Parser::new_ext(
+        text,
+        Options::ENABLE_TABLES
+            | Options::ENABLE_FOOTNOTES
+            | Options::ENABLE_STRIKETHROUGH
+            | Options::ENABLE_TASKLISTS
+            | Options::ENABLE_HEADING_ATTRIBUTES,
+    );
+    for event in parser {
+        match event {
+            Event::Start(tag) => {
+                tag_stack.push((pos, tag));
+            }
+            Event::End(end_tag) => {
+                if let Some((start_offset, tag)) = tag_stack.pop() {
+                    if end_tag != tag {
+                        log::warn!("Mismatched markdown tag");
+                        continue;
+                    }
+
+                    if let Tag::CodeBlock(_kind) = &tag {
+                        code_block_indices.push(start_offset..pos);
+                    }
+
+                    add_attribute_for_tag(
+                        &tag,
+                        builder.add_attributes_for_range(start_offset..pos),
+                        style.clone(),
+                    );
+
+                    if should_add_newline_after_tag(&tag) {
+                        builder.push("\n");
+                        pos += 1;
+                    }
+                } else {
+                    // Very minor warning. We got a bit of malformed markdown (according to
+                    // pulldown_cmark, at least)
```

Comment:
```suggestion
                    // Very minor warning. We got a bit of malformed markdown
```

No need for that, I think :)

Reply:
I just removed the entire comment. The pulldown remark was about an error potentially being specific to pulldown, which doesn't try to include all possible fancy markdown variant.